python/lsst/afw/image/multiband.py

```diff
@@ -0,0 +1,346 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+
+__all__ = ["MultibandBase", "MultibandImage"]
+
+from collections import OrderedDict
+
+import numpy as np
+
+from .image.image import ImageF
+from lsst.afw.geom import Point2I, Box2I
+
+
+class MultibandBase(object):
+    """Base class for multiband objects
+
+    The LSST stack has a number of image-like classes that have
+    data in multiple bands that are stored as separate objects.
+    Analyzing the data can be easier using a Multiband object that
+    wraps the underlying data as a single data cube that can be sliced and
+    updated as a single object.
+
+    `MultibandBase` is designed to contain the most important universal
+    methods for initializing, slicing, and extracting common parameters
+    (such as the bounding box or XY0 position) to all of the single band classes.
+    """
+    def __init__(self, singles, slices=None, **kwargs):
+        """Initialize a `MultibandBase` object
+
+        Must be overloaded in derived classes.
+        At a minimum, the derived class must be initialized with
+        a `_singles` attribute that is a list of single band objects,
+        a `_filters` attribute that is a list of filter
+        names, a `_bbox` attribute is a `Box2I`,
+        and a `_xy0` attribute that is a `Point2I`.
+
+        Below are the only required `__init__` parameters for all
+        dervied classes.
+
+        Parameters
+        ----------
+        singles: list, `OrderedDict`, or `MultibandBase`
+            Either a list of single band objects or an `OrderedDict` with
+            filter names as keys and single band objects as values, or
+            an instance of an object that inherits from `MultibandBase`.
+        slices: list
+            Slices used to extract a subset of an image
+            (used when `singles` inherits from `MultibandBase` to
+            create a new object with a slice of the original).
+        """
+        err = "Inherited classes from `MultibandBase` require an `__init__` method"
+        raise NotImplementedError(err)
+
+    @property
+    def filters(self):
+        """List of filter names for the single band objects
+        """
+        return self._filters
+
+    @property
+    def singles(self):
+        """List of afw single band objects
+        """
+        return self._singles
+
+    @property
+    def bbox(self):
```

Comment:
I'm afraid `bbox` and `xy0` don't meet our requirements for properties (https://developer.lsst.io/python/style.html#properties-should-be-used-when-they-behave-like-regular-instance-attributes), because they return mutable objects that can be modified with surprising results.  In other words, you can do
```
obj.bbox.grow(3)
```
or
```
obj.xy0.setX(4)
```
And end up with an inconsistent object.  Our convention would be to just define getters and setters for these, since it's much less likely that someone would expect e.g. `obj.getBBox().grow(3)` to affect `obj`.